File: disjoint-set/Disjoint-set.hs
{-
Haskell implementation of Disjoint-set Data Structure
-}
find::Int->[Int]->Int --find operation of Data Structure
find x xs
	|x==y = x --if parent return
	|otherwise = find y xs --if not parent continue find
	where y=xs!!x

union::Int->Int->[Int]->[Int]  --union operation of Data Structure
union x y xs=replaceNth (find x xs) (find y xs) xs --make  y's representative , parent of x's representive	 


replaceNth::Int->Int->[Int]->[Int] -- replacing the ith element of list (Complexity O(n))
replaceNth n newVal (x:xs)
     | n == 0 = newVal:xs
     | otherwise = x:replaceNth (n-1) newVal xs

main=do putStrLn "Enter Number of elements :"  --main function
	str<-getLine  	 	
	let n=read str   --number of elements in set
	let set=[0..n-1] --Disjoint set declaration and initialisation
	putStrLn "Enter Number of operations :"
	str<-getLine
	let n=read str --number of operations to be performed
	action n set

action 0 _=return() --terminating loop
action n set=do   --simulates for loop using recursion
	putStrLn "Enter sets to do union :"
  	str<-getLine
	let x=read str::Int --first set
	str<-getLine
	let y=read str::Int --second set
	let set1=union x y set --union operation performed
	let set=set1
	putStrLn $ show set --printing disjoint set representation
	putStrLn "Enter element to do find :"  
	str<-getLine
	let x=read str::Int --element to be searched
	putStrLn $ show $ find x set --find operation performed
	action (n-1) set  --recursive call for next iteration

File: disjoint-set/Disjoint-set.cpp
/*
C++ implementation of Disjoint-Set Data Structure
*/
#include<bits/stdc++.h>
using namespace std;
int fnd(int x,int set[])   //The find operation of the Data Structure
{
	if(set[x]==x)
		return x;
	else
		return set[x]=fnd(set[x],set);
}

void unin(int x,int y,int set[]) //union operation of Data Structure
{
	set[fnd(x,set)]=fnd(y,set);
}

int main()
{
	int i,j,n,m,x,y;
	cout<<"Enter the number of elements : ";
	cin>>n;
	int set[n];     //Disjoint Set
	for(i=0;i<n;i++)
		set[i]=i; //Initializing Disjoint-Set
	cout<<"Enter the number of operations : ";
	cin>>m;
	for(i=0;i<m;i++)
	{
		cout<<"Enter the two elements to do union : ";
		cin>>x>>y;
		unin(x,y,set);
		for(j=0;j<n;j++)
			cout<<set[j]<<" ";
		cout<<"\nEnter the element to do find : ";
		cin>>x;
		cout<<fnd(x,set)<<"\n";
	}
}
